### PR TITLE
[DSA] Handle partial-DP padding in Decode page-table transform

### PR DESCRIPTION
**Pull request by HanHan009527**
## Motivation

Partial-DP attention can pad the local query layout while keeping the request page table unpadded. In the GLM-5.2 DSA + MTP decode path, `_pad_topk_indices()` therefore produces trailing `-1` rows to match the padded query shape, but `metadata.page_table_1` still has one row per real local query.

The unfused DSA decode transform currently requires both tensors to have exactly the same number of rows:

```python
assert page_table.shape[0] == topk_indices.shape[0]
```

With TP16 / DP8, the attention TP size is 2. An otherwise valid partial-DP batch can consequently reach the transform as, for example, a 3-row page table and a 4-row top-k tensor whose last row is all `-1`. The current assertion terminates the scheduler before the Triton kernel runs.

This was first observed while capturing the MTP draft-decode CUDA graph. Disabling CUDA graphs reproduced the same assertion on the eager MTP decode path, so the issue is the partial-DP metadata contract rather than graph capture itself.

## Minimal reproduction on `main`

Run this from an SGLang CUDA environment:

```bash
python3 - <<'PY'
import torch

from sglang.kernels.ops.attention.dsa.transform_index import (
    transform_index_page_table_decode_fast,
)

real_rows = 3
padded_rows = 4
context_length = 8192
topk = 2048

page_table = torch.arange(
    real_rows * context_length,
    dtype=torch.int32,
    device="cuda",
).view(real_rows, context_length)
topk_indices = torch.full(
    (padded_rows, topk),
    -1,
    dtype=torch.int64,
    device="cuda",
)
topk_indices[:real_rows] = torch.arange(topk, device="cuda")

result = transform_index_page_table_decode_fast(page_table, topk_indices)
torch.cuda.synchronize()
assert torch.all(result[-1] == -1)
print("partial-DP padding passed")
PY
```

On the base commit this fails at the equal-row assertion. With this change, the three real rows are transformed and the padded row is filled with `-1`.

## Full topology that triggered the issue

The failure was reproduced by the GLM-5.2 decode role below. Run once per node with `NODE_RANK=0/1`; CUDA Graph capture reaches the failing Draft Decode path during startup.

```bash
python3 -m sglang.launch_server \
  --model-path <GLM-5.2-FP8> \
  --served-model-name glm-5.2 \
  --host 0.0.0.0 \
  --port 30220 \
  --disaggregation-mode decode \
  --disaggregation-transfer-backend mooncake \
  --disaggregation-ib-device "mlx5_1,mlx5_2,mlx5_3,mlx5_4" \
  --trust-remote-code \
  --nnodes 2 \
  --node-rank "${NODE_RANK}" \
  --dist-init-addr "${MASTER_ADDR}:21220" \
  --tp-size 16 \
  --dp-size 8 \
  --ep-size 16 \
  --enable-dp-attention \
  --num-reserved-decode-tokens 2000 \
  --mem-fraction-static 0.9 \
  --max-running-requests 80 \
  --max-total-tokens 205000 \
  --context-length 395000 \
  --page-size 64 \
  --moe-dense-tp-size 1 \
  --enable-dp-lm-head \
  --quantization fp8 \
  --moe-a2a-backend deepep \
  --deepep-mode low_latency \
  --cuda-graph-bs 1 4 10 16 \
  --dsa-topk-backend torch \
  --mooncake-ib-device "mlx5_1,mlx5_2,mlx5_3,mlx5_4" \
  --moe-runner-backend deep_gemm \
  --speculative-moe-runner-backend deep_gemm \
  --speculative-algorithm EAGLE \
  --speculative-num-steps 2 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 3 \
  --kv-cache-dtype fp8_e4m3
```

Adding `--disable-cuda-graph` avoids the startup capture but the same assertion occurs when the first affected eager MTP Decode batch is executed.

## Root cause

The page table and top-k tensor intentionally describe different row domains in this case:

- `page_table.shape[0]`: real local queries/requests;
- `topk_indices.shape[0]`: padded local query layout consumed by the attention backend.

The existing transform assumes both domains are always equal. Merely relaxing the Python assertion would be unsafe because the Triton grid is sized from `topk_indices`, and padded programs would then form page-table pointers for rows that do not exist.

## Changes

- Allow `page_table.shape[0] <= topk_indices.shape[0]`, while continuing to reject the reverse mismatch with a clear error.
- Pass the real page-table row count as a runtime Triton argument and include `req_id < page_table_num_rows` in the load mask.
- Fill padded output rows with the existing `-1` sentinel in the same kernel; no concatenation, allocation, or additional clearing kernel is added to the fast path.
- Keep the row count as a runtime argument rather than a `tl.constexpr`, avoiding a full-value specialization for every real batch size.
- Align the reference implementation with the same row contract.
- Add coverage for partial-DP padding with a dirty preallocated output buffer and for the invalid reverse mismatch.

The existing equal-row correctness, stride, large-batch, and 1M-context test cases remain unchanged and cover the previous behavior.

## Validation

Completed locally:

- `git diff --check`
- `python -m compileall -q python/sglang/kernels/ops/attention/dsa/transform_index.py test/registered/kernels/test_dsa_transform_index.py`
- `ruff format --check` on both changed files
- `ruff check --ignore E741` on both changed files (`E741` is pre-existing in this file)

Pending before marking the PR ready:

- Run `python -m pytest -q test/registered/kernels/test_dsa_transform_index.py` on a CUDA runner.
- Capture the test-only RED result on the base commit and the GREEN result on this commit.
- Re-run the GLM-5.2 TP16 / DP8 / EP16 MTP topology in both eager and Full CUDA Graph modes.
- Compare equal-row transform latency before/after for representative Decode batch sizes.

---
### CI States

Latest PR Test (Base): :x: [Run #29724969152](https://github.com/bytedance-iaas/sglang/actions/runs/29724969152)
Latest PR Test (Extra): :x: [Run #29724969058](https://github.com/bytedance-iaas/sglang/actions/runs/29724969058)

### CODE REVIEW

**Comment by HanHan009527** [comment]
Closing this internal fork PR because the intended contribution target is upstream sgl-project/sglang. The replacement cross-repository Draft PR is https://github.com/sgl-project/sglang/pull/31760.
